man/im_clip.3
.TH CONVERSIONS 3 "15 April 1991"
.SH NAME
im_c2amph, im_c2imag, im_c2ps, im_c2real, im_clip2fmt,
im_ri2c \- conversion between image types
.SH SYNOPSIS
#include <vips/vips.h>

int im_c2amph(in, out)
.br
IMAGE *in, *out;

int im_c2rect(in, out)
.br
IMAGE *in, *out;

int im_c2imag(in, out)
.br
IMAGE *in, *out;

int im_c2ps(in, out)
.br
IMAGE *in, *out;

int im_c2real(in, out)
.br
IMAGE *in, *out;

int im_clip2fmt(in, out, ofmt)
.br
IMAGE *in, *out;
.br
int ofmt;

int im_ri2c(in1, in2, out)
.br
IMAGE *in1, *in2, *out;
.SH DESCRIPTION
Each of the above functions converts the image held by the image descriptor in
to another type image and writes the result on the image descriptor out.
Sizes and the number of channels are identical to those of input.  Whenever
required by the conversion, the element values are rounded by calling floor().

.B im_c2amph()
converts a complex (float or double) input to (amplitude, phase), with phase
in degrees.  Output has the same format as input.

.B im_c2rect()
converts a complex (float or double) input in (amplitude, phase) form back to
rectangular coordinates. Again, phase should be expressed in degrees.

.B im_c2ps()
convert a (float or double) complex image to amplitude (float or
double) image.  Complex input (a,b) is mapped to sqrt( a*a+b*b ).

.B im_clip2fmt()
converts the input image to 'ofmt', where ofmt is the new value for
BandFmt. For example, im_clip2fmt(in, out, FMTUSHORT) converts any image to
unsigned short.

The number of
clipped values (if any), are printed in the standard error output. Before
clipping rounding is carried out if necessary. When converting between real
and complex types with these functions, the imaginary part of the output is
set to zero. When converting from complex to real types, the imaginary part is
ignored. Use im_c2ps(), im_c2real() etc. to get other behaviour.

Legacy functions are available, called im_clip(), im_clip2c(), im_clip2d(),
im_clip2f(), im_clip2i(), im_clip2s(), im_clip2ui(), im_clip2us(),
im_clip2cm() and im_clip2dcm() which convert the input image to unsigned char,
signed char, double, float, int, short, unsigned int, unsigned short,
complex and double complex respectively.

.B im_ri2c() 
takes as inputs two non-complex images of equal sizes and number of
channels.  The output is a complex image with the real part coming from in1
and the imaginary part coming from in2.  Output is float complex (FMTCOMPLEX)
only if none of the inputs is double.  If one or both inputs are double the
result is double complex (FMTDPCOMPLEX).

.B im_c2real() 
and 
.B im_c2imag() 
extract the real or the imaginary part of a
complex image respectively.  If input is be float complex or double complex,
then output is float or double respectively.

.SH RETURN VALUE
The functions returns 0 on success and -1 on error.
.SH SEE ALSO
im_scale(3).


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/conversion/im_clip.c
/* @(#) Clip any down to 0-255. Call im_copy if the image is already uchar. 
 * @(#)
 * @(#) int 
 * @(#) im_clip( in, out )
 * @(#) IMAGE *in, *out;
 * @(#)
 * @(#) Returns 0 on success and -1 on error
 * @(#)
 *
 * Author: Nicos Dessipris
 * Written on: 07/03/1991
 * Modified on: 
 * 04/05/1992 JC
 *	- works for char, uchar too
 *	- floating point code removed from integer clip operations
 *	- uses nint() instead of own rounding code
 *	- calculated the number of >255 clips for float/double input
 *	  incorrectly
 *	- rejects complex input correctly now
 * 27/4/93 JC
 *	- adapted to work with partial images
 *	- nint() removed, now just +0.5
 *	- im_warning code removed
 * 30/6/93 JC
 *	- adapted for partial v2
 * 31/8/93 JC
 *	- now detects and prints over/underflows
 * 27/10/93 JC
 *	- unsigned integer clips now faster!
 *	- falls back to im_copy() correctly
 * 5/5/94 JC
 *	- switched to rint()
 * 18/8/94 JC
 *	- now uses evalend callback
 * 9/5/95 JC
 *	- now does complex too
 * 11/7/95 JC
 *	- now uses IM_RINT() macro
 * 10/3/01 JC
 *	- slightly faster and simpler
 *	- generalised to im_clip2fmt(), all other clippers now just call
 *	  this
 * 21/4/04 JC
 *	- now does floor(), not rint() ... you'll need to round yourself
 *	  before calling this if you want round-to-nearest
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <assert.h>
#include <limits.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Global state. Track over/under-flows for all sequences in this.
 */
typedef struct {
	IMAGE *in;		/* Parameters */
	IMAGE *out;
	int ofmt;

	int underflow;		/* Number of underflows */
	int overflow;		/* Number of overflows */
} Clip;

static int
clip_destroy( Clip *clip )
{
	/* Print warnings, if necessary. 
	 */
	if( clip->overflow || clip->underflow ) 
		im_warn( "im_clip", 
			_( "%d underflows and %d overflows detected" ),
			clip->underflow, clip->overflow );

	return( 0 );
}

/* Build a Clip.
 */
static Clip *
clip_new( IMAGE *in, IMAGE *out, int ofmt )
{
	Clip *clip = IM_NEW( out, Clip );

	if( !clip )
		 return( NULL );

	clip->in = in;
	clip->out = out;
	clip->ofmt = ofmt;
	clip->underflow = 0;
	clip->overflow = 0;

	if( im_add_close_callback( out, 
		(im_callback_fn) clip_destroy, clip, NULL ) )
		return( NULL );

	return( clip );
}

/* Our sequence value: the region this sequence is using, and two local stats.
 */
typedef struct {
	REGION *ir;		/* Input region */
	int underflow;		/* Number of underflows */
	int overflow;		/* Number of overflows */
} ClipSequence;

/* Destroy a sequence value.
 */
static int
stop_clip( ClipSequence *seq, IMAGE *in, Clip *clip )
{
	/* Add to global stats.
	 */
	clip->underflow += seq->underflow;
	clip->overflow += seq->overflow;

	/* Junk our region too.
	 */
	if( seq->ir ) {
		im_region_free( seq->ir );
		seq->ir = NULL;
	}

	return( 0 );
}

/* Make a sequence value.
 */
static void *
start_clip( IMAGE *out, IMAGE *in, Clip *clip )
{
	ClipSequence *seq = IM_NEW( out, ClipSequence );
	 
	if( !seq )
		 return( NULL );

	/* Init!
	 */
	seq->ir = NULL;
	seq->overflow = 0;
	seq->underflow = 0;

	if( !(seq->ir = im_region_create( in )) ) 
		return( NULL );

	return( (void *) seq );
}

/* Clip int types to an int type.
 */
#define IM_CLIP_INT_INT( ITYPE, OTYPE, IM_CLIP ) { \
	ITYPE *p = (ITYPE *) IM_REGION_ADDR( ir, le, y );\
	OTYPE *q = (OTYPE *) IM_REGION_ADDR( or, le, y );\
	\
	for( x = 0; x < sz; x++ ) {\
		int t = p[x];\
		\
		IM_CLIP( t, seq );\
		\
		q[x] = t;\
	}\
}

/* Clip float types to an int type.
 */
#define IM_CLIP_FLOAT_INT( ITYPE, OTYPE, IM_CLIP ) { \
	ITYPE *p = (ITYPE *) IM_REGION_ADDR( ir, le, y );\
	OTYPE *q = (OTYPE *) IM_REGION_ADDR( or, le, y );\
	\
	for( x = 0; x < sz; x++ ) {\
		ITYPE v = floor( p[x] );\
		\
		IM_CLIP( v, seq ); \
		\
		q[x] = v;\
	}\
}

/* Clip complex types to an int type. Just take the real part.
 */
#define IM_CLIP_COMPLEX_INT( ITYPE, OTYPE, IM_CLIP ) { \
	ITYPE *p = (ITYPE *) IM_REGION_ADDR( ir, le, y );\
	OTYPE *q = (OTYPE *) IM_REGION_ADDR( or, le, y );\
	\
	for( x = 0; x < sz; x++ ) {\
		ITYPE v = floor( p[0] );\
		p += 2;\
		\
		IM_CLIP( v, seq ); \
		\
		q[x] = v;\
	}\
}

/* Clip non-complex types to a float type.
 */
#define IM_CLIP_REAL_FLOAT( ITYPE, OTYPE ) { \
	ITYPE *p = (ITYPE *) IM_REGION_ADDR( ir, le, y );\
	OTYPE *q = (OTYPE *) IM_REGION_ADDR( or, le, y );\
	\
	for( x = 0; x < sz; x++ )\
		q[x] = p[x];\
}

/* Clip complex types to a float type ... just take real.
 */
#define IM_CLIP_COMPLEX_FLOAT( ITYPE, OTYPE ) { \
	ITYPE *p = (ITYPE *) IM_REGION_ADDR( ir, le, y );\
	OTYPE *q = (OTYPE *) IM_REGION_ADDR( or, le, y );\
	\
	for( x = 0; x < sz; x++ ) {\
		q[x] = p[0];\
		p += 2;\
	}\
}

/* Clip any to a complex type ... set imaginary to zero.
 */
#define IM_CLIP_ANY_COMPLEX( ITYPE, OTYPE ) { \
	ITYPE *p = (ITYPE *) IM_REGION_ADDR( ir, le, y );\
	OTYPE *q = (OTYPE *) IM_REGION_ADDR( or, le, y );\
	\
	for( x = 0; x < sz; x++ ) {\
		q[0] = p[x];\
		q[1] = 0.0;\
		q += 2;\
	}\
}

#define BAND_SWITCH_INNER( ITYPE, INT, FLOAT, COMPLEX ) { \
	switch( clip->out->BandFmt ) { \
	case IM_BANDFMT_UCHAR: \
		INT( ITYPE, unsigned char, IM_CLIP_UCHAR ); \
		break; \
	case IM_BANDFMT_CHAR: \
		INT( ITYPE, signed char, IM_CLIP_CHAR ); \
		break; \
	case IM_BANDFMT_USHORT: \
		INT( ITYPE, unsigned short, IM_CLIP_USHORT ); \
		break; \
	case IM_BANDFMT_SHORT: \
		INT( ITYPE, signed short, IM_CLIP_SHORT ); \
		break; \
	case IM_BANDFMT_UINT: \
		INT( ITYPE, unsigned int, IM_CLIP_NONE ); \
		break; \
	case IM_BANDFMT_INT: \
		INT( ITYPE, signed int, IM_CLIP_NONE ); \
		break; \
	case IM_BANDFMT_FLOAT: \
		FLOAT( ITYPE, float ); \
		break; \
	case IM_BANDFMT_DOUBLE: \
		FLOAT( ITYPE, double ); \
		break; \
	case IM_BANDFMT_COMPLEX: \
		COMPLEX( ITYPE, float ); \
		break; \
	case IM_BANDFMT_DPCOMPLEX: \
		COMPLEX( ITYPE, double ); \
		break; \
	default: \
		assert( 0 ); \
	} \
}

/* Clip a small area.
 */
static int
clip_gen( REGION *or, ClipSequence *seq, IMAGE *in, Clip *clip )
{	
	REGION *ir = seq->ir;
	Rect *r = &or->valid;
	int le = r->left;
	int to = r->top;
	int bo = IM_RECT_BOTTOM(r);
	int sz = IM_REGION_N_ELEMENTS( or );
	int x, y;

	if( im_prepare( ir, r ) )
		return( -1 );

	for( y = to; y < bo; y++ ) {
		switch( clip->in->BandFmt ) { 
		case IM_BANDFMT_UCHAR: 
			BAND_SWITCH_INNER( unsigned char,
				IM_CLIP_INT_INT, IM_CLIP_REAL_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_CHAR: 
			BAND_SWITCH_INNER( signed char,
				IM_CLIP_INT_INT, IM_CLIP_REAL_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_USHORT: 
			BAND_SWITCH_INNER( unsigned short,
				IM_CLIP_INT_INT, IM_CLIP_REAL_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_SHORT: 
			BAND_SWITCH_INNER( signed short,
				IM_CLIP_INT_INT, IM_CLIP_REAL_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_UINT: 
			BAND_SWITCH_INNER( unsigned int,
				IM_CLIP_INT_INT, IM_CLIP_REAL_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_INT: 
			BAND_SWITCH_INNER( signed int,
				IM_CLIP_INT_INT, IM_CLIP_REAL_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_FLOAT: 
			BAND_SWITCH_INNER( float,
				IM_CLIP_FLOAT_INT, IM_CLIP_REAL_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_DOUBLE: 
			BAND_SWITCH_INNER( double,
				IM_CLIP_FLOAT_INT, IM_CLIP_REAL_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_COMPLEX: 
			BAND_SWITCH_INNER( float,
				IM_CLIP_COMPLEX_INT, IM_CLIP_COMPLEX_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		case IM_BANDFMT_DPCOMPLEX: 
			BAND_SWITCH_INNER( double,
				IM_CLIP_COMPLEX_INT, IM_CLIP_COMPLEX_FLOAT, 
				IM_CLIP_ANY_COMPLEX );
			break; 
		default: 
			assert( 0 ); 
		} 
	}

	return( 0 );
}

/* Clip to any format.
 */
int 
im_clip2fmt( IMAGE *in, IMAGE *out, int ofmt ) 
{
	Clip *clip;

	/* Check args.
	 */
        if( im_piocheck( in, out ) )
		return( -1 );
	if( in->Coding != IM_CODING_NONE ) {
		im_error( "im_clip2fmt", _( "in must be uncoded" ) );
		return( -1 );
	}
	if( ofmt < 0 || ofmt > IM_BANDFMT_DPCOMPLEX ) {
		im_error( "im_clip2fmt", _( "ofmt out of range" ) );
		return( -1 );
	}

	/* Trivial case: fall back to im_copy().
	 */
	if( in->BandFmt == ofmt )
		return( im_copy( in, out ) );

	if( !(clip = clip_new( in, out, ofmt )) )
		return( -1 );

	/* Prepare output header.
	 */
	if( im_cp_desc( out, in ) )
		return( -1 );
	out->BandFmt = ofmt;
	out->Bbits = im_bits_of_fmt( ofmt );

	if( im_demand_hint( out, IM_THINSTRIP, in, NULL ) ||
		im_generate( out, start_clip, clip_gen, stop_clip, in, clip ) )
		return( -1 );

	return( 0 );
}

/* Legacy clippers.
 */
int
im_clip( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_UCHAR ) );
}

int
im_clip2c( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_CHAR ) );
}

int
im_clip2us( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_USHORT ) );
}

int
im_clip2s( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_SHORT ) );
}

int
im_clip2ui( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_UINT ) );
}

int
im_clip2i( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_INT ) );
}

int
im_clip2f( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_FLOAT ) );
}

int
im_clip2d( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_DOUBLE ) );
}

int
im_clip2cm( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_COMPLEX ) );
}

int
im_clip2dcm( IMAGE *in, IMAGE *out )
{
	return( im_clip2fmt( in, out, IM_BANDFMT_DPCOMPLEX ) );
}
